Add userpass and LDAP support for HashiCorp vault credential_plugin

djyasin opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Modify vault credential_plugin to authenticate to vault using LDAP
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 23.6.1
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

djyasin committed:
Add username and password to handle_auth and update exception message

Revise naming of ldap username and password

relrod commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -59,6 +59,13 @@
+            'label': _('LDAP Auth'),
+            'type': 'string',
+            'secret': True,
+            'help_text': _('Username and password for LDAP user authentication.' ' see https://developer.hashicorp.com/vault/docs/auth/ldap'),
```
```suggestion
            'help_text': _('Username and password for LDAP user authentication. See https://developer.hashicorp.com/vault/docs/auth/ldap'),
```

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -59,6 +59,13 @@
                 ' see https://www.vaultproject.io/docs/auth/kubernetes#configuration'
             ),
         },
+        {
+            'id': 'ldap_auth',
+            'label': _('LDAP Auth'),
+            'type': 'string',
+            'secret': True,
+            'help_text': _('Username and password for LDAP user authentication. See https://developer.hashicorp.com/vault/docs/auth/ldap'),
+        },
```
username and password should be 2 separate fields

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -59,6 +59,20 @@
+            'id': 'ldap_username',
+            'label': _('LDAP Auth'),
+            'type': 'string',
+            'secret': True,
```
```suggestion
            'secret': False,
```

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -59,6 +59,20 @@
             ),
         },
+        {
+            'id': 'ldap_username',
```
just username this will be generically use for other username/password auths

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -59,6 +59,20 @@
+            'help_text': _('Username for LDAP user authentication. See https://developer.hashicorp.com/vault/docs/auth/ldap'),
+        },
+        {
+            'id': 'ldap_password',
```
just `password`

djyasin committed:
Add url for LDAP and userpass to method_auth

relrod commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -177,7 +177,7 @@ def handle_auth(**kwargs):
         token = method_auth(**kwargs, auth_param=kubernetes_auth(**kwargs))
-    elif kwargs.get('username') and kwargs.get('password'):
+    elif kwargs.get('username') or kwargs.get('password'):
         token = method_auth(**kwargs, auth_param=ldap_auth(**kwargs))
```
With this change to `or` - what if only one is given? (specifically when `ldap_auth()` is called)

djyasin replied:
Thank you, Rick! I actually hadn't meant to leave this as `or` I was testing to see if it had an impact on the bad gateway error I keep getting, which it did not. I will get this changed!

djyasin committed:
Add information regarding LDAP and username and password to credential plugins documentation

Revise ldap_auth to userpass_auth and revised exception to better reflect functionality

djyasin commented:
Testing steps:

Once your environment with HashiVault and an LDAP user has been configured, this can be tested via UI or CLI. 

- In the AWX UI, navigate to "credentials." 
- From there, click "Add"
- Enter a name for your credential, then choose "HashiCorp Vault Secret Lookup"
- For "Server URL" enter `http://tools_vault_1:1234`
- Enter `awx_ldap_admin` for "Username" and `admin123` for "Password"
- For "Path to Auth" enter `ldap`
- Click "Save"
- Add another credential
- Enter a name for your credential
- Under "Credential Type" choose `Machine`
- Click the key icon by "Username"
- Choose the "HashiVault Secret Lookup" credential name you created earlier
- For "Path to Secret" enter `my_engine/my_root/my_folder`
- For "Key Name" enter `my_key`

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -185,21 +199,25 @@
     else:
-        raise Exception('Either a token or AppRole, Kubernetes, or TLS authentication parameters must be set')
-
+        raise Exception('Token, LDAP, AppRole, Kubernetes, or TLS authentication parameters must be set')
```
```suggestion
        raise Exception('Token, Username/Password, AppRole, Kubernetes, or TLS authentication parameters must be set')
```

TheRealHaoLiu commented on awx/main/credential_plugins/hashivault.py:
```diff
@@ -185,21 +199,25 @@
     return token
 
 
+def ldap_auth(**kwargs):
```
should be named something else since this is being use for any username/password auth

djyasin committed:
Revise method_auth to ensure certs can be used with username and ensure namespace functionality is not hindered

jjwatt reviewed (approved):
LGTM, Approval.
Tested against devel with instructions from AAP-17485
